[Core/UX] Add Job API and support managed job (both on-demand and spot)

Michaelvll opened the pull request:
<!-- Describe the changes in this PR -->

~~Note: this PR require a `sky start -f sky-spot-controller-xx` to make the commands like `sky spot logs`, `sky spot queue` work.~~ (We found a way to handle this)


Refer to the latest doc: https://skypilot.readthedocs.io/en/job-api/examples/managed-jobs.html#using-both-spot-and-on-demand-instances

TODO:
  - [x] Update the managed spot doc
  - [x] Add doc for pipeline
  - [x] Better hints when `sky job queue` runs on an old controller
  - [x] Test backward compatibility

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
  - [x] `pytest tests/test_smoke.py`
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`
  - [x] master: `sky spot launch -n pipeline tests/test_yamls/pipeline.yaml`; this PR: `sky job queue`; `sky job logs`; `sky job dashboard`; `sky job cancel`; `sky job queue`
  - [x] `bash -i tests/backward_comaptibility_tests.sh 0 7`

cblmemo commented on sky/utils/schemas.py:
```diff
@@ -5,40 +5,130 @@
 
 
-def get_single_resources_schema():
+def _resources_properties_without_accelerators():
```
why dont include accs?

cblmemo reviewed (commented):
Thanks for refactoring!! This is huge 🚀🚀🚀 Just finished a quick round of review. Could you resolve the merge conflict? Also, there is still some `spot` hiding in comments, READMEs or code; could you double-check as well?

Michaelvll commented on sky/job/core.py:
```diff
@@ -53,45 +53,44 @@ def launch(
-            raise ValueError('Only single-task or chain DAG is allowed for '
-                             f'sky.spot.launch. Dag:\n{dag}')
+    assert dag.is_chain(), ('Only single-task or chain DAG is '
+                            'allowed for job_launch.', dag)
```
Good point! Changed to error

Michaelvll commented on sky/job/core.py:
```diff
@@ -53,45 +53,44 @@ def launch(
-                    'Either change task names to be unique, or specify the DAG '
-                    'name only and comment out the task names (so that they '
-                    'will be auto-generated) .')
+            raise ValueError(
```
Good catch! Added back.

Michaelvll commented on sky/job/core.py:
```diff
@@ -53,45 +53,44 @@ def launch(
-                                     mode='w') as f:
+    with tempfile.NamedTemporaryFile(prefix=f'managed-dag-{dag.name}-',
+                                     mode='w',
+                                     delete=False) as f:
```
Oops, it was for debugging purpose. Added back. Thanks!

Michaelvll commented on sky/job/utils.py:
```diff
@@ -36,7 +36,9 @@
 # Add user hash so that two users don't have the same controller VM on
 # shared-account clouds such as GCP.
-SPOT_CONTROLLER_NAME: str = (
+JOB_CONTROLLER_NAME: str = (
+    f'sky-job-controller-{common_utils.get_user_hash()}')
+LEGACY_JOB_CONTROLLER_NAME: str = (
```
It seems adding significant overhead for the controller. Also, many clouds do not support cloning a VM at the moment. It might be fine to just ask the user to delete the old controller themselves.

Michaelvll commented on sky/resources.py:
```diff
@@ -352,7 +353,12 @@ def cpus(self) -> Optional[str]:
+        if self.cloud is not None and self._instance_type is not None:
+            vcpus, _ = self.cloud.get_vcpus_mem_from_instance_type(
+                self._instance_type)
+            return str(vcpus)
```
it is to align with the `accelerators` to always return the actual cpus when instance type is specified. This helps the printing of the resources in the `sky job launch`.

Michaelvll commented on sky/utils/controller_utils.py:
```diff
@@ -48,74 +49,95 @@
+    @property
+    def cluster_name(self) -> str:
+        """Returns the cluster name that exists."""
+        for candidate_name in self.candidate_cluster_names:
+            record = global_user_state.get_cluster_from_name(candidate_name)
+            if record is not None:
+                return candidate_name
```
No, it is not possible, because:
1. if the old controller exist: we will always reuse the old one without creating a new job-controller
2. If the old controller does not exist: we will create a new one and always use the new job controller.

Michaelvll commented on sky/utils/env_options.py:
```diff
@@ -11,7 +11,7 @@ class Options(enum.Enum):
     # Internal: this is used to skip the cloud user identity check,
     # which is used to protect cluster operations in a multi-identity
-    # scenario. Currently, this is only used in the spot controller,
+    # scenario. Currently, this is only used in the job controller,
```
Good catch! Fixed the comment. Thanks!

cblmemo commented on sky/job/controller.py:
```diff
@@ -292,22 +294,23 @@ def _run_one_task(self, task_id: int, task: 'sky.Task') -> bool:
                 resources = handle.launched_resources
                 assert resources is not None, handle
-                if resources.need_cleanup_after_preemption():
+                if (resources.use_spot and
+                        resources.need_cleanup_after_preemption()):
```
Should we do this for ondemand as well? iiuc it also needs cleanup after an unexpected error

concretevitamin replied:
Kinda have the same feeling. If future resources.need_cleanup_after_preemption() impl added an on-demand case, this code won't automatically handle it?

cblmemo commented on sky/job/controller.py:
```diff
@@ -370,27 +373,27 @@ def run(self):
+                                    callback_func=callback_func)
 
 
 def _run_controller(job_id: int, dag_yaml: str, retry_until_up: bool):
```
```suggestion
def _run_controller(controller_job_id: int, dag_yaml: str, retry_until_up: bool):
```
should we rename it to reduce the confusion w/ managed job id?

cblmemo replied:
SG!

cblmemo commented on sky/job/controller.py:
```diff
@@ -43,8 +43,8 @@ def _get_dag_and_name(dag_yaml: str) -> Tuple['sky.Dag', str]:
 
-class SpotController:
-    """Each spot controller manages the life cycle of one spot job."""
+class JobController:
```
```suggestion
class ManagedJobController:
```
not sure

cblmemo commented on sky/job/controller.py:
```diff
@@ -409,28 +412,29 @@ def _handle_signal(job_id):
+        task is executed at a time.
     """
     # NOTE: The code to get cluster name is same as what we did in the spot
     # controller, we should keep it in sync with SpotController.__init__()
```
```suggestion
    # controller, we should keep it in sync with JobController.__init__()
```

cblmemo commented on sky/job/controller.py:
```diff
@@ -457,16 +461,16 @@ def start(job_id, dag_yaml, retry_until_up):
+    except exceptions.ManagedJobUserCancelledError:
         dag, _ = _get_dag_and_name(dag_yaml)
-        task_id, _ = (spot_state.get_latest_task_id_status(job_id))
+        task_id, _ = (job_state.get_latest_task_id_status(job_id))
```
```suggestion
        task_id, _ = job_state.get_latest_task_id_status(job_id)
```

cblmemo commented on sky/job/controller.py:
```diff
@@ -480,37 +484,38 @@ def start(job_id, dag_yaml, retry_until_up):
-        # directly, and then cleanup the cluster state.
+        # directly, and then cleanup the cluster job_state.
         _cleanup(job_id, dag_yaml=dag_yaml)
         logger.info(f'Spot cluster of job {job_id} has been cleaned up.')
```
```suggestion
        logger.info(f'Worker cluster of job {job_id} has been cleaned up.')
```

cblmemo commented on sky/job/core.py:
```diff
@@ -14,9 +14,9 @@
 from sky import task as task_lib
 from sky.backends import backend_utils
 from sky.clouds.service_catalog import common as service_catalog_common
+from sky.job import constants
+from sky.job import utils
```
```suggestion
from sky.job import constants as job_constants
from sky.job import utils as job_utils
```

cblmemo commented on sky/job/dashboard/dashboard.py:
```diff
@@ -12,36 +12,39 @@
             pathlib.Path('~/.sky/sky_ray.yml').expanduser().read_text())
-        return config.get('cluster_name', '').startswith('sky-spot-controller-')
+        cluster_name = config.get('cluster_name', '')
+        return (cluster_name.startswith('sky-spot-controller-') or
+                cluster_name.startswith('sky-job-controller-'))
```
Could we use the `Controllers.candidate_name` instead?

cblmemo commented on sky/job/dashboard/dashboard.py:
```diff
@@ -12,36 +12,39 @@
 
-import sky
-from sky import spot
+from sky import job
```
```suggestion
from sky import job as managed job
```
nit

concretevitamin replied:
+1, similar feeling. Left a similar comment elsewhere. This particular instance and some other instances seem too generic (e.g., job_utils).

cblmemo commented on sky/job/recovery_strategy.py:
```diff
@@ -17,8 +17,8 @@
 from sky import sky_logging
 from sky import status_lib
 from sky.backends import backend_utils
+from sky.job import utils
```
```suggestion
from sky.job import utils as job_utils
```

cblmemo commented on sky/job/state.py:
```diff
@@ -124,21 +126,21 @@
 
-class SpotStatus(enum.Enum):
+class ManagedJobStatus(enum.Enum):
     """Spot job status, designed to be in serverless style.
```
```suggestion
    """Managed job status, designed to be in serverless style.
```

cblmemo commented on sky/job/state.py:
```diff
@@ -147,37 +149,37 @@ class SpotStatus(enum.Enum):
         FAILED          ->  FAILED
         FAILED_SETUP    ->  FAILED_SETUP
     Note that the JobStatus will not be stuck in PENDING, because each spot
```
```suggestion
    Note that the JobStatus will not be stuck in PENDING, because each worker
```

cblmemo commented on sky/job/utils.py:
```diff
@@ -18,11 +18,11 @@
 from sky import sky_logging
 from sky.backends import backend_utils
+from sky.job import constants as job_constants
+from sky.job import state
```
```suggestion
from sky.job import state as job_state
```

cblmemo commented on sky/job/utils.py:
```diff
@@ -663,46 +698,46 @@ def format_job_table(
 
     Usage:
 
       >> codegen = SpotCodegen.show_jobs(...)
```
```suggestion
      >> codegen = ManagedJobCodeGen.show_jobs(...)
```

cblmemo commented on sky/skylet/events.py:
```diff
@@ -14,10 +14,10 @@
 from sky import sky_logging
 from sky.backends import cloud_vm_ray_backend
 from sky.clouds import cloud_registry
+from sky.job import utils
```
```suggestion
from sky.job import utils as job_utils
```

cblmemo commented on sky/job/utils.py:
```diff
@@ -663,46 +698,46 @@ def format_job_table(
     _PREFIX = [
-        'from sky.spot import spot_state',
-        'from sky.spot import spot_utils',
+        'from sky.job import state',
+        'from sky.job import utils',
```
```suggestion
        'from sky.job import state as job_state',
        'from sky.job import utils as job_utils',
```

cblmemo commented on sky/utils/controller_utils.py:
```diff
@@ -20,11 +20,11 @@
 from sky.data import data_utils
 from sky.data import storage as storage_lib
+from sky.job import constants as job_constants
+from sky.job import utils
```
```suggestion
from sky.job import utils as job_utils
```

cblmemo commented on sky/utils/schemas.py:
```diff
@@ -58,9 +82,14 @@ def _get_single_resources_schema():
             'use_spot': {
                 'type': 'boolean',
             },
+            # Deprecated: use 'job_recovery' instead. This is for backward
```
should we print out some warnings for deprecated config field?

cblmemo commented on sky/utils/controller_utils.py:
```diff
@@ -299,18 +322,22 @@ def get_controller_resources(
         # config.
         custom_controller_resources_config = skypilot_config.get_nested(
-            (controller_type, 'controller', 'resources'), None)
+            (controller.value.controller_type, 'controller', 'resources'), None)
```
seems like the `controller_type` is `jobs`, while the config schema says `managed_jobs`. align those two?

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -14,20 +14,20 @@
 from sky import cli
 from sky import exceptions
 from sky import global_user_state
+from sky import job
```
```suggestion
from sky import job as managed_job
```

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -236,10 +236,10 @@ def test_stop_spot_controller(self, _mock_cluster_state,
+                f'\'{job.JOB_CONTROLLER_NAME}\' is currently not supported'
                 in result.output)
 
         result = cli_runner.invoke(cli.autostop, ['sky-spot-con*'])
```
```suggestion
        result = cli_runner.invoke(cli.autostop, ['sky-job-con*'])
```

cblmemo commented on tests/test_smoke.py:
```diff
@@ -13,7 +13,7 @@
 #
-# Only run managed spot tests
+# Only run managed jobtests
 # > pytest tests/test_smoke.py --managed-spot
```
```suggestion
# > pytest tests/test_smoke.py --managed-jobs
```

Michaelvll replied:
Fixed. Thanks!

cblmemo reviewed (commented):
More comments..! Mostly nits

concretevitamin commented on docs/source/reference/yaml-spec.rst:
```diff
@@ -92,10 +92,16 @@ Available fields:
-      # `use_spot` must be True for this to have any effect. For now, only
-      # `FAILOVER` strategy is supported.
-      spot_recovery: none
+      # The recovery strategy for managed jobs (optional).
```
For discussion: How about we hide this field since the default EAGER_NEXT_REGION has been working well?

concretevitamin commented on docs/source/reference/yaml-spec.rst:
```diff
@@ -92,10 +92,16 @@ Available fields:
-      # `use_spot` must be True for this to have any effect. For now, only
-      # `FAILOVER` strategy is supported.
-      spot_recovery: none
+      # The recovery strategy for managed jobs (optional).
+      # For now, only `FAILOVER` and `EAGER_NEXT_REGION` strategy is supported.
+      # If `FAILOVER` is specified, the job will be restarted in the same region
+      # if the node fails, and go to the next region if no available resources
+      # are found in the same region. If `EAGER_NEXT_REGION` is specified, the
+      # job will go to the next region directly if the node fails. This is
+      # useful for spot instances, as in practice, preemptions in a region
+      # usually indicate a shortage of resources in that region.
+      # For managed jobs, default: EAGER_NEXT_REGION
+      job_recovery: none
```
```suggestion
      # The recovery strategy for managed jobs (optional).
      #
      # In effect for managed jobs. Possible values are `FAILOVER` and `EAGER_NEXT_REGION`.
      #
      # If `FAILOVER` is specified, the job will be restarted in the same region
      # if the node fails, and go to the next region if no available resources
      # are found in the same region. 
      #
      # If `EAGER_NEXT_REGION` is specified, the
      # job will go to the next region directly if the node fails. This is
      # useful for spot instances, as in practice, preemptions in a region
      # usually indicate a shortage of resources in that region.
      #
      # , default: EAGER_NEXT_REGION
      job_recovery: none
```
Paragraph spacing (important for readability & consistency across repo).

concretevitamin commented on sky/job/__init__.py:
```diff
@@ -0,0 +1,43 @@
+"""Modules for managed job clusters."""
```
```suggestion
"""Managed Jobs."""
```

concretevitamin commented on sky/job/constants.py:
```diff
@@ -1,16 +1,16 @@
-"""Constants used for Managed Spot."""
+"""Constants used for Managed Job."""
```
```suggestion
"""Constants used for Managed Jobs."""
```

Michaelvll replied:
Fixed. Thanks!

concretevitamin commented on sky/resources.py:
```diff
@@ -731,30 +738,30 @@ def _try_validate_cpus_mem(self) -> None:
                             f'{self.instance_type} does not have the requested '
                             f'number of vCPUs. {self.instance_type} has {cpus} '
-                            f'vCPUs, but {self.cpus} is requested.')
+                            f'vCPUs, but {self._cpus} is requested.')
```
The _cpus change makes these messages (and above) unnecessary? Also, a bit inaccurate too since it's possible _cpus is not explicitly "requested". What's the best way to fix?

concretevitamin commented on sky/resources.py:
```diff
@@ -770,25 +777,20 @@ def _try_validate_cpus_mem(self) -> None:
 
-    def _try_validate_spot(self) -> None:
-        """Try to validate the spot related attributes.
+    def _try_validate_job(self) -> None:
+        """Try to validate the job related attributes.
```
```suggestion
    def _try_validate_managed_job_attributes(self) -> None:
        """Try to validate the managed job related attributes.
```

concretevitamin commented on sky/resources.py:
```diff
@@ -1290,7 +1292,14 @@ def _from_yaml_config_single(cls, config: Dict[str, str]) -> 'Resources':
-        resources_fields['spot_recovery'] = config.pop('spot_recovery', None)
+        if config.get('spot_recovery') is not None:
+            logger.warning('spot_recovery is deprecated. Use job_recovery '
+                           'instead.')
```
```suggestion
                           'instead (the system is defaulting to that for you).')
```

Michaelvll replied:
Updated. Thanks!

concretevitamin commented on sky/resources.py:
```diff
@@ -1379,9 +1388,6 @@ def __setstate__(self, state):
         if version < 2:
             self._region = None
 
-        if version < 3:
-            self._spot_recovery = None
```
Is this necessary? If not, better to keep it in history?

concretevitamin replied:
Just to make the history of field versions monotonic. It's more readable; otherwise, readers are left to guess what happened in this gap.

Michaelvll replied:
nit: How about we comment out the version 3 check and add a comment here? Still, just to keep the code being executed is relevant. In the future, we may even want a policy that removes some of these backward compatibility checks as what we did for the other places (remove backward compatibility code after 2/3 major releases).

concretevitamin replied:
I still think we should just keep the history. Purging old fields is a good idea, but we should do it in chunks, rather than leaving "holes".

For example, with it commented I notice that switching from this PR to the master shows
```
» sky spot launch exit 1 -y
...

Launching managed spot job 'sky-cmd' from spot controller...
Launching spot controller...
AttributeError: 'Resources' object has no attribute '_spot_recovery'
```

concretevitamin commented on tests/test_smoke.py:
```diff
@@ -13,7 +13,7 @@
 # > pytest tests/test_smoke.py::test_minimal
 #
-# Only run managed spot tests
+# Only run managed jobtests
```
```suggestion
# Only run managed jobtests
```
```suggestion
# Only run managed job tests
```

Michaelvll replied:
Fixed. Thanks!

concretevitamin commented on sky/utils/controller_utils.py:
```diff
@@ -52,78 +52,99 @@ class _ControllerSpec:
 
+    @property
+    def cluster_name(self) -> str:
+        """Returns the cluster name that exists."""
```
```suggestion
        """Returns the name in candidate_cluster_names that exists, else the first name."""
```

Michaelvll replied:
Updated. Thanks!

concretevitamin commented on sky/cli.py:
```diff
@@ -3539,12 +3566,32 @@ def spot_dashboard(port: Optional[int]):
+    """Alias for Managed Job CLI (default to spot instances)."""
+    pass
+
+
```
Also, for this group of aliases, the following message
> WARNING: `sky spot queue` is deprecated and will be removed in a future release. Please use `sky job queue` instead.

isn't entirely accurate. How about something like

> NOTE: `sky spot queue` has been renamed to `sky job queue` and will be removed in a future release. Please use the latter instead.

Michaelvll replied:
Sounds good. Updated the message, but I would rather keep `WARNING` instead to be more alerting for user to update the command used.

concretevitamin commented on sky/execution.py:
```diff
@@ -160,11 +160,11 @@ def _execute(
             raise ValueError(
-                'Spot recovery is specified in the task. To launch the '
-                'managed spot job, please use: sky spot launch')
+                'Job recovery is specified in the task. To launch the '
+                'managed job, please use: sky job launch')
```
```suggestion
                'Job recovery is specified in the task. To launch it as a '
                'managed job, please use: sky job launch')
```

Michaelvll replied:
Fixed. Thanks!

Michaelvll commented on docs/source/examples/managed-jobs.rst:
```diff
@@ -377,13 +330,83 @@ To submit the pipeline, the same command :code:`sky spot launch` is used. The pi
+
+1. Use a lower-cost controller (if you have a low number of concurrent managed jobs).
+2. Enforcing the job controller to run on a specific location. (Default: cheapest location)
+3. Changing the maximum number of jobs that can be run concurrently, which is 2x the vCPUs of the controller. (Default: 16)
```
Good point! updated. Thanks!

concretevitamin commented on sky/job/core.py:
```diff
@@ -1,4 +1,4 @@
-"""SDK functions for managed spot job."""
+"""SDK functions for managed job."""
```
```suggestion
"""SDK functions for managed jobs."""
```

Michaelvll replied:
Updated. Thanks!

concretevitamin commented on sky/job/core.py:
```diff
@@ -33,20 +33,21 @@ def launch(
     stream_logs: bool = True,
     detach_run: bool = False,
     retry_until_up: bool = False,
-) -> None:
```
Keep? Always show -> None rather than omit?

Michaelvll replied:
Oops, this is a merging error. Fixed. Thanks!

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -2557,41 +2556,58 @@ def get_task_demands_dict(task: 'task_lib.Task') -> Dict[str, float]:
 
 
-def get_task_resources_str(task: 'task_lib.Task') -> str:
+def get_task_resources_str(task: 'task_lib.Task',
```
Are changes here supposed to display in `sky job queue`? See my UX comment. I had an old autostopped controller, then upgraded to the this PR and ran 2 `job launch`.

Michaelvll replied:
Yes, this is for `sky job queue`. All the existing spot jobs will not have the `(spot)` annotation in the job table, while all newly launched jobs will have them, if it is on an existing controller. Do you think it is necessary to add `(spot)` for those old jobs? I found it may need a few hacky string processing code, which I would rather avoid.

concretevitamin commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -3181,33 +3182,34 @@ def _exec_code_on_head(
                     f'{backend_utils.RESET_BOLD}'
-                    '\nTo view the spot job dashboard:\t'
-                    f'{backend_utils.BOLD}sky spot dashboard'
+                    '\nTo view the managed job dashboard:\t'
+                    f'{backend_utils.BOLD}sky job dashboard'
```
See ux comment.

Michaelvll replied:
Fixed. Thanks!

concretevitamin commented on sky/job/core.py:
```diff
@@ -33,20 +33,21 @@ def launch(
+    """Launch a managed job.
 
-    Please refer to the sky.cli.spot_launch for the document.
+    Please refer to the sky.cli.job_launch for the document.
```
```suggestion
    Please refer to sky.cli.job_launch for documentation.
```
(document != documentation)

Michaelvll replied:
Done. THanks!

concretevitamin commented on sky/job/core.py:
```diff
@@ -134,9 +136,9 @@ def launch(
+    """Get statuses of managed jobs.
 
-    Please refer to the sky.cli.spot_queue for the documentation.
+    Please refer to the sky.cli.job_queue for the documentation.
```
```suggestion
    Please refer to sky.cli.job_queue for documentation.
```

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/job/core.py:
```diff
@@ -222,18 +225,18 @@ def cancel(name: Optional[str] = None,
+    """Cancel managed jobs.
 
-    Please refer to the sky.cli.spot_cancel for the document.
+    Please refer to the sky.cli.job_cancel for the document.
```
```suggestion
    Please refer to sky.cli.job_cancel for documentation.
```

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/job/dashboard/dashboard.py:
```diff
@@ -12,36 +12,47 @@
+    if not _is_running_on_job_controller():
         # Experimental: run on laptop (refresh is very slow).
-        all_spot_jobs = sky.spot_queue(refresh=True, skip_finished=False)
+        all_managed_jobs = job.queue(refresh=True, skip_finished=False)
```
managed_jobs.* seems better.

Michaelvll replied:
Fixed. Thanks!

concretevitamin commented on sky/job/dashboard/templates/index.html:
```diff
@@ -107,6 +106,32 @@ <h1>Managed spot jobs</h1>
         </table>
     </div>
+    <script>
+        document.querySelectorAll('.folder').forEach(function (folder) {
```
Add comment: is folding for pipelines only?

Michaelvll replied:
Added. Thanks!

concretevitamin commented on sky/job/recovery_strategy.py:
```diff
@@ -1,9 +1,9 @@
-"""The strategy to handle launching/recovery/termination of spot clusters.
+"""The strategy to handle launching/recovery/termination of clusters.
```
```suggestion
"""The strategy to handle launching/recovery/termination of managed job clusters.
```

Michaelvll replied:
Fixed. Thanks!

concretevitamin commented on sky/job/recovery_strategy.py:
```diff
@@ -50,17 +50,17 @@ def terminate_cluster(cluster_name: str, max_retry: int = 3) -> None:
 
 class StrategyExecutor:
-    """Handle each launching, recovery and termination of the spot clusters."""
+    """Handle each launching, recovery and termination of the clusters."""
```
```suggestion
    """Handle the launching, recovery and termination of managed job clusters."""
```

Michaelvll replied:
Fixed.

concretevitamin commented on sky/job/recovery_strategy.py:
```diff
@@ -214,8 +214,8 @@ def _wait_until_job_starts_on_cluster(self) -> Optional[float]:
             try:
-                status = spot_utils.get_job_status(self.backend,
-                                                   self.cluster_name)
+                status = job_utils.get_job_status(self.backend,
```
managed_job_utils

Michaelvll replied:
ok, fixed. Thanks!

concretevitamin commented on sky/job/recovery_strategy.py:
```diff
@@ -330,12 +330,12 @@ def _launch(self,
                             reasons=reasons)
                     return None
-                logger.info('Failed to launch the spot cluster with error: '
+                logger.info('Failed to launch the cluster with error: '
```
(Since cluster failed to launch it doesn't exist so more accurate to omit "the")
```suggestion
                logger.info('Failed to launch cluster with error: '
```

Michaelvll replied:
I suppose we should at least add an `a` there to replace `the`.

concretevitamin commented on sky/job/recovery_strategy.py:
```diff
@@ -330,12 +330,12 @@ def _launch(self,
                 # If the launch fails, it will be recovered by the following
                 # code.
-                logger.info('Failed to launch the spot cluster with error: '
+                logger.info('Failed to launch the cluster with error: '
```
```suggestion
                logger.info('Failed to launch cluster with error: '
```

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/job/recovery_strategy.py:
```diff
@@ -358,13 +358,13 @@ def _launch(self,
-                        raise exceptions.SpotJobReachedMaxRetriesError(
-                            'Resources unavailable: failed to launch the spot '
+                        raise exceptions.ManagedJobReachedMaxRetriesError(
+                            'Resources unavailable: failed to launch the '
```
```suggestion
                            'Resources unavailable: failed to launch '
```

Michaelvll replied:
Fixed. Thanks!

concretevitamin commented on docs/source/docs/index.rst:
```diff
@@ -139,7 +139,7 @@ Contents
    :caption: Cutting Cloud Costs
 
-   ../examples/spot-jobs
+   ../examples/managed-jobs
```
Also for brainstorming: this may be more indirect than before, where people could clearly see "ah managed spot is clearly something I can use to save costs".

concretevitamin reviewed (commented):
Awesome @Michaelvll, this is a major addition to the API!
 
52 / 61 files viewed for now, mostly sky/spot/* left.

----

```
  job          Managed Job CLI (jobs with auto-recovery).
  spot         Alias for Managed Job CLI (default to spot instances).
```

Leaving this for brainstorming. Kind of feel like we should find better descriptions for these.

-----

On `sky job queue`
```

ID  TASK  NAME       RESOURCES    SUBMITTED   TOT. DURATION  JOB DURATION  #RECOVERIES  STATUS
3   -     sky-cmd    1x[CPU:2+]   1 min ago   1m 8s          -             0            STARTING
2   -     sky-cmd    1x[CPU:2+]   4 mins ago  2m 49s         6s            0            FAILED`
```
one is on-demand, the other spot. Have we decided that the RESOURCES column shouldn't show on-demand vs. spot?

---

UX:
```

I 04-28 15:33:43 cloud_vm_ray_backend.py:3196] Managed Job ID: 3
I 04-28 15:33:43 cloud_vm_ray_backend.py:3196] To cancel the job:               sky job cancel 3
I 04-28 15:33:43 cloud_vm_ray_backend.py:3196] To stream job logs:              sky job logs 3
I 04-28 15:33:43 cloud_vm_ray_backend.py:3196] To stream controller logs:       sky job logs --controller 3
I 04-28 15:33:43 cloud_vm_ray_backend.py:3196] To view all managed jobs:        sky job queue
I 04-28 15:33:43 cloud_vm_ray_backend.py:3196] To view the managed job dashboard:       sky job dashboard
```
Last line is unaligned.

Michaelvll commented on docs/source/docs/index.rst:
```diff
@@ -121,7 +121,7 @@ Contents
    :caption: Running Jobs
 
-   ../examples/spot-jobs
+   ../examples/managed-jobs
```
Good point! Added. Thanks!

Michaelvll commented on sky/skylet/events.py:
```diff
@@ -14,10 +14,10 @@
 from sky import sky_logging
 from sky.backends import cloud_vm_ray_backend
 from sky.clouds import cloud_registry
+from sky.job import utils as job_utils
```
I would prefer to keep the module name to be `utils`, but we can always import the module with name `managed_job_utils` if necessary. Reason: `job/managed_job_utils.py` is quite verbose and duplicated, and this is to keep align with the other modules such as `provision/gcp/instance.py` instead of `provision/gcp/gcp_instance.py`

concretevitamin replied:
> can always import the module with name `managed_job_utils` if necessary.

This sounds good.

Michaelvll commented on sky/utils/schemas.py:
```diff
@@ -5,6 +5,30 @@
 """
 
 
+def _only_one_of_two_field_config(field1: str, field2: str):
```
Updated. Thanks!

Michaelvll commented on tests/backward_compatibility_tests.sh:
```diff
@@ -158,35 +158,37 @@ sky logs ${CLUSTER_NAME}-6 2 --status
-SPOT_JOB_JOB_NAME=${CLUSTER_NAME}-${uuid:0:4}
+MANAGED_JOB_JOB_NAME=${CLUSTER_NAME}-${uuid:0:4}
+# Test managed jobs to make sure existing jobs and new job can run correctly, after
+# the job controller is updated.
```
Updated. Thanks!

Michaelvll commented on tests/conftest.py:
```diff
@@ -18,8 +18,8 @@
 #
-# To only run tests for managed spot (without generic tests), use
-# --managed-spot.
+# To only run tests for managed job (without generic tests), use
```
Updated. Thanks!

Michaelvll commented on tests/test_jobs_and_serve.py:
```diff
@@ -131,18 +131,18 @@ def _mock_serve_controller(_mock_db_conn):
 
 class TestSpotOperations:
-    """Test operations for managed spot."""
+    """Test operations for managed job."""
```
Do you mean renaming the `job/` folder to `managed_job` or `managed_jobs`? It seems `ray` uses `job` (non-plural) for module name: https://github.com/ray-project/ray/tree/master/dashboard/modules/job

concretevitamin replied:
I meant in docstrs, we should refer to the functionality/module as "managed jobs" (plural).

Michaelvll commented on sky/cli.py:
```diff
@@ -3539,12 +3566,32 @@ def spot_dashboard(port: Optional[int]):
+    pass
+
+
+_add_command_alias(job,
```
Added for CLI. Thanks!

concretevitamin replied:
```
» sky spot launch exit 1 --use-spot                                       1 ↵
WARNING: `sky spot launch` has been renamed to `sky job launch` and will be removed in a future release. Instead, please use the latter with additional arguments --use_spot.
```
Last arg use_spot -> use-spot

Michaelvll commented on sky/exceptions.py:
```diff
@@ -52,9 +52,9 @@ class InvalidCloudConfigs(Exception):
 
 class ProvisionPrechecksError(Exception):
-    """Raised when a spot job fails prechecks before provision.
+    """Raised when a managed job fails prechecks before provision.
```
Fixed. Thanks!

Michaelvll commented on sky/job/README.md:
```diff
@@ -1,6 +1,6 @@
+# SkyPilot Managed Jobs
 
-This module is used for running user jobs on spot clusters, which automatically recovers the job from preemptions.
+This module is used for running and managing user jobs, which automatically recovers the job from spot preemptions and machine failures.
```
Thanks! Fixed.

Michaelvll commented on sky/cli.py:
```diff
@@ -830,17 +830,43 @@ def wrapper(self, *args, **kwargs):
+def _override_arguments(callback, override_command_argument: Dict[str, Any]):
+
+    def wrapper(*args, **kwargs):
+        logger.info(f'Overriding arguments: {override_command_argument}')
```
It is used in `_add_command_alias` function?

concretevitamin replied:
Oh, I meant is this supposed to be user-facing?

Michaelvll replied:
Ah, yes, I am trying to keep this logging to let the user know `sky spot launch` overrides the argument `--use-spot`. Wdyt?

concretevitamin commented on sky/cli.py:
```diff
@@ -816,31 +816,74 @@ def get_help(self, ctx):
-            f'WARNING: `{alias_name}` is deprecated and will be removed in a '
-            f'future release. Please use `{original_name}` instead.\n',
+            f'WARNING: `{alias_name}` has been renamed to `{original_name}` '
+            f'and will be removed in a future release. Instead, please use the '
+            f'latter{override_str}.\n',
```
```suggestion
            f'and will be removed in a future release. Please use the '
            f'latter{override_str} instead.\n',
```

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -2557,41 +2556,58 @@ def get_task_demands_dict(task: 'task_lib.Task') -> Dict[str, float]:
+                task_cpu_demand = f'{min_cpus}+'
+            if 'Spot' in spot_type:
+                spot_str = '|'.join(sorted(spot_type))
+                spot_str = f'({spot_str})'
```
Nit: [Spot] reads better? Also in line with `status` outputs, e.g., 1x GCP(n2-standard-8[Spot]).

Michaelvll replied:
ok, updated to `[]` now.

concretevitamin reviewed (commented):
1. We should refer to this module as "Managed Jobs" (plural). Here & elsewhere:
```
  job          Managed Job CLI (jobs with auto-recovery).
  spot         Alias for Managed Job CLI (default to managed spot jobs).
```

concretevitamin commented on docs/source/reference/yaml-spec.rst:
```diff
@@ -92,10 +92,21 @@ Available fields:
+      #
+      # In effect for managed jobs. Possible values are `FAILOVER` and `EAGER_NEXT_REGION`.
+      #
+      # If `FAILOVER` is specified, the job will be restarted in the same region
+      # if the node fails, and go to the next region if no available resources
+      # are found in the same region. 
+      #
+      # If `EAGER_NEXT_REGION` is specified, the
+      # job will go to the next region directly if the node fails. This is
+      # useful for spot instances, as in practice, preemptions in a region
+      # usually indicate a shortage of resources in that region.
```
nit: format each para

concretevitamin commented on sky/job/utils.py:
```diff
@@ -164,16 +167,17 @@ def get_job_timestamp(backend: 'backends.CloudVmRayBackend', cluster_name: str,
+            task.name, job_id) if task.name else None
+        logger.info(f'=== START: event callback for {status!r} ===')
+        log_path = os.path.join(constants.SKY_LOGS_DIRECTORY,
+                                'managed_jobs_event',
```
```suggestion
                                'managed_job_event',
```
Should we use this to be consistent with a bunch of imports?

concretevitamin commented on sky/job/utils.py:
```diff
@@ -72,7 +73,7 @@ class UserSignal(enum.Enum):
 def get_job_status(backend: 'backends.CloudVmRayBackend',
                    cluster_name: str) -> Optional['job_lib.JobStatus']:
-    """Check the status of the job running on the spot cluster.
+    """Check the status of the job running on the cluster.
```
```suggestion
    """Check the status of the job running on a managed job cluster.
```

concretevitamin commented on sky/job/utils.py:
```diff
@@ -299,43 +308,46 @@ def stream_logs_by_id(job_id: int, follow: bool = True) -> str:
+        # task_id and managed_job_status can be None if the controller process
+        # just started and the managed job status has not set to PENDING yet.
+        while managed_job_status is None or not managed_job_status.is_terminal(
+        ):
```
nit: wrap the whole clause in ()?

concretevitamin commented on sky/job/state.py:
```diff
@@ -124,60 +126,59 @@
-    Each spot job submitted to the spot cluster, will have a JobStatus
-    on that spot cluster:
+    The ManagedJobStatus is a higher level status than the JobStatus.
+    Each managed job submitted to a cluster, will have a JobStatus
+    on that cluster:
```
```suggestion
    Each managed job submitted to a cluster will have a JobStatus
    on that cluster:
```

concretevitamin commented on sky/job/core.py:
```diff
@@ -222,18 +227,18 @@ def cancel(name: Optional[str] = None,
+    """Cancel managed jobs.
 
-    Please refer to the sky.cli.spot_cancel for the document.
+    Please refer to the sky.cli.job_cancel for the documentation.
```
```suggestion
    Please refer to sky.cli.job_cancel for documentation.
```

concretevitamin commented on sky/job/core.py:
```diff
@@ -276,42 +281,48 @@ def cancel(name: Optional[str] = None,
+    """Tail logs of managed jobs.
 
-    Please refer to the sky.cli.spot_logs for the document.
+    Please refer to the sky.cli.job_logs for the document.
```
```suggestion
    Please refer to sky.cli.job_logs for documentation.
```

concretevitamin commented on sky/job/controller.py:
```diff
@@ -249,7 +249,8 @@ def _run_one_task(self, task_id: int, task: 'sky.Task') -> bool:
                 logger.info(
-                    f'Cluster is preempted{cluster_status_str}. Recovering...')
+                    f'Cluster is preempted or failed{cluster_status_str}. '
+                    'Recovering...')
```
Update all comments in this func w.r.t. "preempted"? To account for the new cases.

Also, shall we test an on-demand managed job's cluster failing + recovering?

Michaelvll replied:
I tested the manual termination and it works correctly as we did for spot preemptions.

For the simulation of a maintenance event, it seems a CPU instance will not experience any downtime as GCP will migrate the whole instance, which works correctly with our managed job.

For the GPU instance, it seems the `gcloud compute instances simulate-maintenance-event` will hang forever, though no obvious reason could be found online.

concretevitamin commented on sky/cli.py:
```diff
@@ -3181,8 +3225,8 @@ def spot():
         'cluster is up, if unavailability errors are encountered. This '  # pylint: disable=bad-docstring-quotes
-        'applies to launching the spot clusters (both the initial and any '
-        'recovery attempts), not the spot controller.'))
+        'applies to launching the managed job clusters (both the initial and '
```
```suggestion
        'applies to launching all managed jobs (both the initial and '
```

concretevitamin reviewed (commented):
Thanks @Michaelvll! Read everything except docs.

---

1. (unrelated?) UX: `sky spot launch tests/test_yamls/pipeline.yaml  --use-spot`

On 0.4, task name display looks good:
```

I 04-29 21:09:28 optimizer.py:760] Best plan:
I 04-29 21:09:28 optimizer.py:765] -----------------------------------------------------------------------------------------------
I 04-29 21:09:28 optimizer.py:765]  TASK    #NODES   CLOUD   INSTANCE              vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE
I 04-29 21:09:28 optimizer.py:765] -----------------------------------------------------------------------------------------------
I 04-29 21:09:28 optimizer.py:765]  a       1        GCP     n2-standard-2[Spot]   2       8         -              us-east4-a
I 04-29 21:09:28 optimizer.py:765]  b       1        GCP     n2-standard-2[Spot]   2       8         -              us-east4-a
I 04-29 21:09:28 optimizer.py:765]  eval1   1        GCP     n2-standard-2[Spot]   2       8         -              us-east4-a
I 04-29 21:09:28 optimizer.py:765]  eval2   1        GCP     n2-standard-2[Spot]   2       8         -              us-east4-a
I 04-29 21:09:28 optimizer.py:765] -----------------------------------------------------------------------------------------------
```

This branch & master
```

I 04-29 21:10:30 optimizer.py:821] ---------------------------------------------------------------------------------------------------------------------------------------------------
I 04-29 21:10:30 optimizer.py:821]  TASK                                                        #NODES   CLOUD   INSTANCE              vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE
I 04-29 21:10:30 optimizer.py:821] ---------------------------------------------------------------------------------------------------------------------------------------------------
I 04-29 21:10:30 optimizer.py:821]  Task<name=a>(run='echo train starts\ns...')                 1        GCP     n2-standard-2[Spot]   2       8         -              us-east4-a
I 04-29 21:10:30 optimizer.py:821]    resources: {GCP([Spot], cpus=2+), AWS([Spot], cpus=2+)}
I 04-29 21:10:30 optimizer.py:821]  Task<name=b>(run='echo train starts\ns...')                 1        AWS     m6i.large             2       8         -              us-east-1
I 04-29 21:10:30 optimizer.py:821]    resources: {AWS(cpus=2+), GCP(cpus=2+)}
I 04-29 21:10:30 optimizer.py:821]  Task<name=eval1>(run='echo eval starts\nsl...')             1        AWS     m6i.large             2       8         -              us-east-1
I 04-29 21:10:30 optimizer.py:821]    resources: <Cloud>(cpus=2+)
I 04-29 21:10:30 optimizer.py:821]  Task<name=eval2>(run='echo eval starts\nsl...')             1        AWS     m6i.large             2       8         -              us-east-1
I 04-29 21:10:30 optimizer.py:821]    resources: <Cloud>(cpus=2+)
I 04-29 21:10:30 optimizer.py:821] ---------------------------------------------------------------------------------------------------------------------------------------------------
```

2. UI: 
- `-7`: I read it as a negative number. Maybe we can put a dropdown thingy in the ID field, but after the number? Clearly shown as a clickable thing.
- STARTING at top is not stylized as a button. Is it intended?
- The autorefresh would reset the expanded/collapsed state. Is it expected?

<img width="1337" alt="Screen Shot 2024-04-29 at 21 15 48" src="https://github.com/skypilot-org/skypilot/assets/592670/de4e90b6-b795-439a-b3a5-a2a232d77afc">

cblmemo commented on sky/backends/backend_utils.py:
```diff
@@ -2557,41 +2556,58 @@ def get_task_demands_dict(task: 'task_lib.Task') -> Dict[str, float]:
         accelerator_dict = task.best_resources.accelerators
+        if is_managed_job:
+            if task.best_resources.use_spot:
+                spot_str = '(spot)'
```
Oh ic! That makes sense. Sounds good

cblmemo commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -3569,17 +3571,17 @@ def tail_logs(self,
-                       job_id: Optional[int] = None,
-                       job_name: Optional[str] = None,
-                       follow: bool = True) -> None:
+    def tail_managed_job_logs(self,
```
I mean the job id for any `sky exec`

cblmemo commented on sky/job/controller.py:
```diff
@@ -88,23 +88,23 @@ def __init__(self, job_id: int, dag_yaml: str,
         """
-        spot_job_logs_dir = os.path.join(constants.SKY_LOGS_DIRECTORY,
-                                         'spot_jobs')
+        managed_job_logs_dir = os.path.join(constants.SKY_LOGS_DIRECTORY,
+                                            'managed_jobs')
```
SOunds good!

Michaelvll commented:
Thanks for the comments @concretevitamin! 
Fixed the optimizer table:

```
sky job launch -n pipeline tests/test_yamls/pipeline.yaml
I 04-30 19:24:21 optimizer.py:816] Best plan: 
I 04-30 19:24:21 optimizer.py:821] ------------------------------------------------------------------------------------------------------------------
I 04-30 19:24:21 optimizer.py:821]  TASK    #NODES   CLOUD        INSTANCE              vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE                 
I 04-30 19:24:21 optimizer.py:821] ------------------------------------------------------------------------------------------------------------------
I 04-30 19:24:21 optimizer.py:821]  a       1        GCP          n2-standard-2[Spot]   2       8         -              northamerica-northeast2-a   
I 04-30 19:24:21 optimizer.py:821]  b       1        AWS          m6i.large             2       8         -              us-east-1                   
I 04-30 19:24:21 optimizer.py:821]  eval1   1        Kubernetes   2CPU--2GB             2       2         -              kubernetes                  
I 04-30 19:24:21 optimizer.py:821]  eval2   1        Kubernetes   2CPU--2GB             2       2         -              kubernetes                  
I 04-30 19:24:21 optimizer.py:821] ------------------------------------------------------------------------------------------------------------------
I 04-30 19:24:21 optimizer.py:821] 
```


Fixed the UI:

<img width="1363" alt="image" src="https://github.com/skypilot-org/skypilot/assets/6753189/1a526cb1-6142-4310-bf82-0d696e648ecd">

concretevitamin commented on sky/job/dashboard/dashboard.py:
```diff
@@ -53,11 +53,12 @@ def home():
     rows = managed_jobs.format_job_table(all_managed_jobs,
                                          show_all=True,
                                          return_rows=True)
+    rows = [[''] + row for row in rows]
```
Add a comment on what the 1st column is.

Michaelvll replied:
Added. Thanks!

concretevitamin commented on sky/job/utils.py:
```diff
@@ -709,76 +710,79 @@ class ManagedJobCodeGen:
+    _PREFIX = textwrap.dedent("""\
+        managed_job_version = 0
+        try:
+            from sky.job import constants, state, utils;
+            managed_job_version = constants.MANAGED_JOB_VERSION;
```
`;` needed?

Michaelvll replied:
Good catch! Fixed. Thanks!

concretevitamin commented on sky/job/utils.py:
```diff
@@ -709,76 +710,79 @@ class ManagedJobCodeGen:
-        ]
+        code = textwrap.dedent(f"""\
+            state.set_job_name({job_id}, {dag_name!r})
+            
```
```suggestion
```

concretevitamin commented on sky/resources.py:
```diff
@@ -1389,6 +1389,11 @@ def __setstate__(self, state):
 
+        # spot_recovery is deprecated. We do not need to set the _spot_recovery
+        # here any more.
+        # if version < 3:
+        #     self._spot_recovery = None
```
I was proposing we keep this uncommented in the history. Any downsides?

Michaelvll replied:
Please see: https://github.com/skypilot-org/skypilot/pull/3419#discussion_r1584103885 : )

Mostly to keep the codes executed are actually needed. : )

concretevitamin commented on sky/job/constants.py:
```diff
@@ -20,3 +20,8 @@
+
+# The version of the lib files that job/utils use. Whenever there is an API
+# change for the job/utils, we need to bump this version, so that the
+# user can be notified to update their SkyPilot version on the remote cluster.
```
What does "user can be notified to update their SkyPilot version on the remote cluster." mean? Notified automatically by the system (where?), or by human devs?

Michaelvll replied:
Ahh, that is a remnant. Now fixed.

concretevitamin reviewed (commented):
Sending non-doc related comments for now.

---

Q: is the PR description 
> Note: this PR require a sky start -f sky-spot-controller-xx to make the commands like sky spot logs, sky spot queue work.

accurate, now that we've added more back compat handling?

----

UI: 
> The autorefresh would reset the expanded/collapsed state. Is it expected?

This problem still exists? Can we use local storage handling just as what we did for `Auto-refresh (every 30s)` to memoize the dropdown's for pipelines?

-----

UI: (feel free to leave to the future) Also, after clicking a dropdown, the whole table's width changes, a bit weird.

----

> Fixed the optimizer table:

`sky job launch -n pipeline tests/test_yamls/pipeline.yaml` seems to still print some weird repr's
```
...
I 05-01 09:06:35 optimizer.py:842] Considered resources for task "Task<name=eval2>(run='echo eval starts\\nsl...')\n  resources: <Cloud>(cpus=2+)" (1 node):
...
```

concretevitamin commented:
Another problem: 
- upgraded to c78077a8c5
- ran some `job launch`
- after a while the controller autostopped
- now ran `status -r`
```
...
sky-spot-controller-8a3968f2   41 mins ago  1x GCP(n2-standard-2, disk_size=50)                       STOPPED  10m       sky spot launch tests/tes...
...
Managed jobs
Failed to query managed jobs due to connection issues. Try again later. Details: [RuntimeError] Command $([ -s ~/.sky/python_path ] && cat ~/.sky/python_path 2> /dev/null || which python3) -u -c 'managed_job_version = 0
try:
    from sky.job import constants, state, utils;
    managed_job_version = constants.MANAGED_JOB_VERSION;
except ImportError:
    from sky.spot import spot_state as state, spot_utils as utils

if managed_job_version < 1:
    job_table = utils.dump_spot_job_queue()
else:
    job_table = utils.dump_managed_job_queue()
print(job_table, flush=True)
' failed with return code 255.
Failed to fetch managed jobs

Services
No live services.

concretevitamin commented:
Problem: `--use-spot` not recognized

```
» sky job launch --use-spot false                                    1 ↵
Task from command: false
Managed job 'sky-cmd' will be launched on (estimated):
I 05-01 10:08:38 optimizer.py:694] == Optimizer ==
I 05-01 10:08:38 optimizer.py:705] Target: minimizing cost
I 05-01 10:08:38 optimizer.py:717] Estimated cost: $0.1 / hour
I 05-01 10:08:38 optimizer.py:717]
I 05-01 10:08:38 optimizer.py:842] Considered resources (1 node):
I 05-01 10:08:38 optimizer.py:912] ------------------------------------------------------------------------------------------------------
I 05-01 10:08:38 optimizer.py:912]  CLOUD   INSTANCE              vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE       COST ($)   CHOSEN
I 05-01 10:08:38 optimizer.py:912] ------------------------------------------------------------------------------------------------------
I 05-01 10:08:38 optimizer.py:912]  GCP     n2-standard-8[Spot]   8       32        -              us-east4-a        0.09          ✔
I 05-01 10:08:38 optimizer.py:912]  AWS     m6i.2xlarge[Spot]     8       32        -              ap-northeast-3a   0.12
I 05-01 10:08:38 optimizer.py:912] ------------------------------------------------------------------------------------------------------
I 05-01 10:08:38 optimizer.py:912]
Launching a managed job 'sky-cmd'. Proceed? [Y/n]:
```

Michaelvll commented:
> accurate, now that we've added more back compat handling?

Good catch! Fixed the description.


> This problem still exists? Can we use local storage handling just as what we did for Auto-refresh (every 30s) to memoize the dropdown's for pipelines?

Oops, it was accidentally deleted in a commit. Restored it now. 

> sky job launch -n pipeline tests/test_yamls/pipeline.yaml seems to still print some weird repr's

Oops, fixed. Thanks for catching this!

> Problem: --use-spot not recognized

`--use-spot false` should be `--no-use-spot`?

concretevitamin commented:
Problem:
```
sky spot queue -r
WARNING: `sky spot queue` has been renamed to `sky job queue` and will be removed in a future release. Please use the latter instead.

Fetching managed job statuses...
Managed jobs
Failed to query managed jobs: [ValueError] Cluster 'sky-job-controller-8a3968f2' does not exist.
```

concretevitamin commented on docs/source/docs/index.rst:
```diff
@@ -121,7 +121,7 @@ Contents
    :caption: Running Jobs
 
-   ../examples/spot-jobs
+   ../examples/managed-jobs
```
Add "new"?

concretevitamin commented on docs/source/reference/cli.rst:
```diff
@@ -59,6 +56,30 @@ Job Queue CLI
+   :nested: full
+
+
 Sky Serve CLI
```
```suggestion
SkyServe CLI
```

concretevitamin reviewed (commented):
Pushed some edits to docs, PTAL.

Michaelvll commented on sky/clouds/gcp.py:
```diff
@@ -841,9 +841,9 @@ def get_current_user_identity_str(cls) -> Optional[str]:
-        """Returns whether a spot resource needs cleanup after preeemption."""
+    def need_cleanup_after_preemption_or_failure(
+            self, resources: 'resources.Resources') -> bool:
+        """Whether a resource needs cleanup after preeemption or failure."""
```
This is currently only for the TPU, and IIRC, on-demand TPU pod could fail and may require cleanup after the failure.

concretevitamin replied:
Should we update L847-850?

Michaelvll commented on sky/job/constants.py:
```diff
@@ -20,3 +20,8 @@
 # long and truncated twice during the cluster creation.
 JOB_CLUSTER_NAME_PREFIX_LENGTH = 25
+
+# The version of the lib files that job/utils use. Whenever there is an API
```
Added. Thanks!

Michaelvll commented on sky/templates/job-controller.yaml.j2:
```diff
@@ -15,17 +15,17 @@ setup: |
-  # Start the controller for the current spot job.
-  python -u -m sky.spot.controller {{remote_user_yaml_path}} \
+  # Start the controller for the current managed job.
+  python -u -m sky.job.controller {{remote_user_yaml_path}} \
```
Tested with `bash -i ./tests/backward_compatibility_tests.sh 0 7` and it seems working : )

concretevitamin commented on sky/job/constants.py:
```diff
@@ -1,4 +1,8 @@
-"""Constants used for Managed Jobs."""
+"""Constants used for Managed Jobs.
+
+NOTE: whenever an API change is made to job/utils, we need to bump the
```
I meant this should be added to job/utils.

concretevitamin commented on sky/job/constants.py:
```diff
@@ -22,6 +26,6 @@
 # The version of the lib files that job/utils use. Whenever there is an API
-# change for the job/utils, we need to bump this version, so that the
-# user can be notified to update their SkyPilot version on the remote cluster.
+# change for the job/utils, we need to bump this version and update the
```
```suggestion
# change for the job/utils, we need to bump this version and update
```

concretevitamin reviewed (commented):
Some final comments. Awesome work @Michaelvll!

cblmemo commented on sky/utils/schemas.py:
```diff
@@ -58,9 +82,14 @@ def _get_single_resources_schema():
                 'type': 'boolean',
             },
+            # Deprecated: use 'job_recovery' instead. This is for backward
+            # compatibility, and can be removed in 0.7.0.
```
```suggestion
            # compatibility, and can be removed in 0.8.0.
```
we decided to deprecate it in 0.8.0 iiuc?

cblmemo commented on sky/utils/schemas.py:
```diff
@@ -619,8 +650,11 @@ def get_config_schema():
         'required': [],
         'additionalProperties': False,
         'properties': {
+            'jobs': controller_resources_schema,
```
should this be jobs or job? should we keep it align with the cli

Michaelvll replied:
We now changed all to `jobs` now to keep CLI, API, and config aligned.

cblmemo replied:
sounds good

cblmemo commented on sky/cli.py:
```diff
@@ -778,8 +778,8 @@ def _make_task_or_dag_from_entrypoint_with_overrides(
         task.workdir = workdir
 
     # Spot launch specific.
```
```suggestion
    # Managed jobs launch specific.
```

cblmemo commented on sky/cli.py:
```diff
@@ -3248,7 +3323,7 @@ def spot_launch(
-        spot_recovery=spot_recovery,
+        job_recovery=job_recovery,
     )
     # Deprecation.
```
what version does we deprecate this? Could we remove it now

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -97,16 +97,16 @@ def _mock_cluster_state(_mock_db_conn):
 @pytest.fixture
 def _mock_spot_controller(_mock_db_conn):
```
```suggestion
def _mock_job_controller(_mock_db_conn):
```

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -97,16 +97,16 @@ def _mock_cluster_state(_mock_db_conn):
-        cluster_name_on_cloud=spot.SPOT_CONTROLLER_NAME,
+        cluster_name=job.JOB_CONTROLLER_NAME,
+        cluster_name_on_cloud=job.JOB_CONTROLLER_NAME,
         cluster_yaml='/tmp/spot_controller.yaml',
```
```suggestion
        cluster_yaml='/tmp/job_controller.yaml',
```

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -131,18 +131,18 @@ def _mock_serve_controller(_mock_db_conn):
+    """Test operations for managed jobs."""
 
     @pytest.mark.timeout(60)
     def test_down_spot_controller(self, _mock_cluster_state,
```
```suggestion
    def test_down_job_controller(self, _mock_cluster_state,
```

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -131,18 +131,18 @@ def _mock_serve_controller(_mock_db_conn):
     return record['handle']
 
 
 class TestSpotOperations:
```
```suggestion
class TestJobOperations:
```

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -183,19 +183,19 @@ def mock_get_job_table_one_job(cls, handle, code, require_outputs,
-        assert 'WARNING: Tearing down the managed spot controller.' in result.output, (
+        assert 'WARNING: Tearing down the managed job controller.' in result.output, (
             result.exception, result.output, result.exc_info)
         assert isinstance(result.exception, exceptions.NotSupportedError)
```
L202, 206, `sky-spot-con`

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -219,9 +219,9 @@ def mock_get_job_table_one_job(cls, handle, code, require_outputs,
     def test_stop_spot_controller(self, _mock_cluster_state,
```
```suggestion
    def test_stop_job_controller(self, _mock_cluster_state,
```

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -219,9 +219,9 @@ def mock_get_job_table_one_job(cls, handle, code, require_outputs,
+        assert (f'Stopping controller(s) \'{job.JOB_CONTROLLER_NAME}\' is '
                 'currently not supported' in result.output)
 
         result = cli_runner.invoke(cli.stop, ['sky-spot-con*'])
```
```suggestion
        result = cli_runner.invoke(cli.stop, ['sky-job-con*'])
```

cblmemo commented on tests/test_jobs_and_serve.py:
Seems like there are lot of `spot` in this file. could you double-check?

cblmemo commented on tests/test_smoke.py:
```diff
@@ -2135,38 +2135,38 @@ def test_stop_gcp_spot():
 @pytest.mark.no_kubernetes  # Kubernetes does not have a notion of spot instances
-@pytest.mark.managed_spot
+@pytest.mark.managed_jobs
 def test_spot(generic_cloud: str):
```
```suggestion
def test_job(generic_cloud: str):
```

cblmemo commented on tests/test_smoke.py:
spotting a lot of `spot` in this file as well 👀

cblmemo commented on sky/utils/controller_utils.py:
```diff
@@ -52,78 +52,100 @@ class _ControllerSpec:
+            record = global_user_state.get_cluster_from_name(candidate_name)
+            if record is not None:
+                return candidate_name
+        return self.candidate_cluster_names[0]
```
add a comment says we should put the latest name in the first element of the list?

cblmemo commented on sky/utils/controller_utils.py:
```diff
@@ -180,7 +203,7 @@ def _get_cloud_dependencies_installation_commands(
     ]
     # k8s and ibm doesn't support open port and spot instance yet, so we don't
     # install them for either controller.
```
Feeling like for now we could install dependencies for every cloud on the job controller. every cloud should support at least ondemand instance and should support managed ondemand jobs at least?

cblmemo commented on sky/job/utils.py:
```diff
@@ -248,6 +255,9 @@ def cancel_jobs_by_id(job_ids: Optional[List[int]]) -> str:
                 f.write(UserSignal.CANCEL.value)
                 f.flush()
+            # Backward compatibility for managed jobs launched before #3419. It
+            # can be removed in the future 0.7.0 release.
```
```suggestion
            # can be removed in the future 0.8.0 release.
```

cblmemo reviewed (commented):
Thanks for the quick fix @Michaelvll ! More comments, mostly nits ;)

Michaelvll commented on docs/source/examples/managed-jobs.rst:
```diff
@@ -0,0 +1,465 @@
+Job YAML
+~~~~~~~~
+
+To launch a managed job, you can simply reuse your job YAML (recommended to test it with :code:`sky launch` first).
```
Let's call it job instead for consistency

Michaelvll commented:
Tested:
- [ ] `pytest tests/test_smoke.py --managed-jobs` (b855c66)

Michaelvll commented on docs/source/images/job-dashboard.png:
updated the figure. Thanks!

cblmemo commented on docs/source/examples/managed-jobs.rst:
```diff
@@ -0,0 +1,465 @@
+    echo eval model on test set
+
+
+The YAML above defines a pipeline with two tasks. The first :code:`name:
+pipeline` names the pipeline. The first task has name :code:`train` and the
+second task has name :code:`eval`. The tasks are separated by a line with three
+dashes :code:`---`. Each task has its own :code:`resources`, :code:`setup`, and
+:code:`run` sections. Tasks are executed sequentially.
```
Sounds good!

cblmemo commented on sky/utils/controller_utils.py:
```diff
@@ -160,40 +184,80 @@ def from_type(cls, controller_type: str) -> Optional['Controllers']:
-            'pip list | grep azure-cli > /dev/null 2>&1 || '
-            'pip install azure-cli>=2.31.0 azure-core azure-identity>=1.13.0 '
-            'azure-mgmt-network > /dev/null 2>&1')
+    commands = []
+    prefix_str = 'Check & install cloud dependencies on controller: '
+    # This is to make sure the shorter checking message does not have junk
+    # characters from the previous message.
+    empty_str = ' ' * 5
+    aws_dependencies_installation = (
+        'pip list | grep boto3 > /dev/null 2>&1 || '
+        'pip install "urllib3<2" awscli>=1.27.10 botocore>=1.29.10 '
+        'boto3>=1.26.1 "colorama<0.4.5" > /dev/null 2>&1')
+    for cloud in sky_check.get_cached_enabled_clouds_or_refresh():
+        if isinstance(
+                clouds,
+            (clouds.Lambda, clouds.SCP, clouds.Fluidstack, clouds.Paperspace)):
+            # no need to install any cloud dependencies for lambda, scp,
+            # fluidstack and paperspace
+            continue
+        if cloud.is_same_cloud(clouds.AWS()):
+            commands.append(f'echo -n "{prefix_str}AWS{empty_str}" && ' +
+                            aws_dependencies_installation)
+        elif cloud.is_same_cloud(clouds.Azure()):
+            commands.append(
+                f'echo -en "\\r{prefix_str}Azure{empty_str}" && '
+                'pip list | grep azure-cli > /dev/null 2>&1 || '
+                'pip install "azure-cli>=2.31.0" azure-core '
+                '"azure-identity>=1.13.0" azure-mgmt-network > /dev/null 2>&1')
+        elif cloud.is_same_cloud(clouds.GCP()):
+            commands.append(
+                f'echo -en "\\r{prefix_str}GCP{empty_str}" && '
+                'pip list | grep google-api-python-client > /dev/null 2>&1 || '
+                'pip install "google-api-python-client>=2.69.0" '
+                '> /dev/null 2>&1')
```
looks great!

cblmemo commented on sky/utils/controller_utils.py:
```diff
@@ -160,40 +184,80 @@ def from_type(cls, controller_type: str) -> Optional['Controllers']:
+                commands.append(f'echo -en "\\r{prefix_str}OCI{empty_str}" && '
+                                'pip list | grep oci > /dev/null 2>&1 || '
+                                'pip install oci > /dev/null 2>&1')
+            elif isinstance(cloud, clouds.Kubernetes):
+                commands.append(
+                    f'echo -en "\\r{prefix_str}Kubernetes{empty_str}" && '
+                    'pip list | grep kubernetes > /dev/null 2>&1 || '
+                    'pip install "kubernetes>=20.0.0" > /dev/null 2>&1')
```
should we install k8s for serve controller as well?

Michaelvll replied:
Seems our serving on kubernetes PR has not merged yet #3377?

cblmemo replied:
Oh got it. Then it looks good to me :)

concretevitamin commented on docs/source/reference/config.rst:
```diff
@@ -14,12 +14,12 @@ Available fields and semantics:
-  # These take effects only when a spot controller does not already exist.
+  # These take effects only when a managed job controller does not already exist.
   #
   # Ref: https://skypilot.readthedocs.io/en/latest/examples/spot-jobs.html#customizing-spot-controller-resources
```
Note: with redirect, tag #customizing-spot-controller-resources is stripped. Is it possible to support? Very minor.

Michaelvll replied:
Yes! Just made the redirection to keep the tags, and also updated the link here.

concretevitamin commented on examples/managed_job.yaml:
I guess the question is whether we want to keep using "managed spot". For discussion. If so, such example files can keep there orig. name.

Michaelvll replied:
Sounds good. I changed this file back to `managed_spot.yaml` and keep the `managed_job_with_storage` as is, so we have both case included in examples.

concretevitamin commented on examples/managed_job_with_storage.yaml:
I guess the question is whether we want to keep using "managed spot". For discussion. If so, such example files can keep there orig. name.

concretevitamin commented on sky/cli.py:
```diff
@@ -3629,15 +3634,27 @@ def spot():
+_add_command_alias(jobs, jobs_dashboard, new_group=spot)
+
+
+@cli.group(cls=_NaturalOrderGroup, hidden=True)
+def job():
+    """Alias for Managed Jobs CLI."""
+    pass
+
+
+_add_command_alias(jobs, jobs_launch, new_group=job, with_warning=False)
+_add_command_alias(jobs, jobs_queue, new_group=job, with_warning=False)
+_add_command_alias(jobs, jobs_logs, new_group=job, with_warning=False)
+_add_command_alias(jobs, jobs_cancel, new_group=job, with_warning=False)
+_add_command_alias(jobs, jobs_dashboard, new_group=job, with_warning=False)
```
I'd prefer to not add the alias group for now. To keep all usage / examples clean, and to avoid the potential confusion?

Michaelvll replied:
I don't have a strong feeling for this. Just removed it for now, though I am a bit used to `sky job launch` (maybe just I am too familiar).

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -503,8 +503,9 @@ def need_cleanup_after_preemption_or_failure(
         no matter the preemption behavior is to terminate or stop the cluster.
-        The only exception by far is GCP's Spot TPU VM. We override this method
-        in gcp.py.
+        Similar for on-demand resources that goes into maintenance mode. The
```
```suggestion
        Similar for on-demand resources that go into maintenance mode. The
```

Michaelvll replied:
Fixed. Thanks!

concretevitamin commented on sky/jobs/utils.py:
```diff
@@ -1,4 +1,10 @@
-"""User interfaces with managed jobs."""
+"""User interfaces with managed jobs.
+
+
```
```suggestion
```

Michaelvll replied:
Fixed. Thanks!

concretevitamin commented on sky/jobs/utils.py:
```diff
@@ -1,4 +1,10 @@
+
+
+NOTE: whenever an API change is made in this file, we need to bump the
+job.constants.MANAGED_JOB_VERSION and handle the API change in the
```
```suggestion
jobs.constants.MANAGED_JOB_VERSION and handle the API change in the
```

Michaelvll replied:
Fixed. Thanks!

concretevitamin reviewed (commented):
A few final things to discuss. Should be good to go in today!

- [ ] This doc needs to change `sky job ..` to `sky jobs ..`? https://skypilot.readthedocs.io/en/job-api/examples/managed-jobs.html

concretevitamin reviewed (approved):
LGTM!

cblmemo commented on sky/jobs/constants.py:
```diff
@@ -1,16 +1,16 @@
 # Based on profiling, memory should be at least 3x (in GB) as num vCPUs to avoid
-# OOM (each vCPU can have 4 spot controller processes as we set the CPU
+# OOM (each vCPU can have 4 job controller processes as we set the CPU
 # requirement to 0.25, and 3 GB is barely enough for 4 spot processes).
```
```suggestion
# requirement to 0.25, and 3 GB is barely enough for 4 job processes).
```

cblmemo commented on sky/jobs/core.py:
```diff
@@ -71,56 +72,59 @@ def launch(
     for task_ in dag.tasks:
         controller_utils.maybe_translate_local_file_mounts_and_sync_up(
-            task_, path='spot')
+            task_, path='job')
```
```suggestion
            task_, path='jobs')
```
should this be jobs

cblmemo commented on sky/resources.py:
```diff
@@ -9,9 +9,9 @@
 from sky import check as sky_check
 from sky import clouds
 from sky import exceptions
+from sky import jobs
```
```suggestion
from sky import jobs as managed_jobs
```

cblmemo commented on sky/serve/constants.py:
```diff
@@ -53,7 +53,7 @@
 CONTROLLER_RESOURCES = {'cpus': '4+', 'disk_size': 200}
 
-# Due to the CPU/memory usage of the controller process launched with sky job (
+# Due to the CPU/memory usage of the controller process launched with sky jobs (
```
accidental? feeling like we should keep the original comment to avoid get confused w/ managed jobs

cblmemo commented on sky/skylet/constants.py:
```diff
@@ -187,11 +187,11 @@
+# Due to the CPU/memory usage of the controller process launched with sky jobs (
+# use ray job under the hood), we need to reserve some CPU/memory for each job/
 # serve controller process.
 # Spot: A default controller with 8 vCPU and 32 GB memory can manage up to 32
```
```suggestion
# Jobs: A default controller with 8 vCPU and 32 GB memory can manage up to 32
```

cblmemo commented on tests/test_jobs_and_serve.py:
```diff
@@ -233,63 +233,62 @@ def test_stop_spot_controller(self, _mock_cluster_state,
     @pytest.mark.timeout(60)
-    def test_down_spot_controller(self, _mock_cluster_state,
-                                  _mock_serve_controller, monkeypatch):
+    def test_down_job_controller(self, _mock_cluster_state,
```
```suggestion
    def test_down_serve_controller(self, _mock_cluster_state,
```
seems like i accidentally copy-paste the function name loll

cblmemo commented on sky/utils/controller_utils.py:
```diff
@@ -299,18 +367,22 @@ def get_controller_resources(
             controller_resources_config_copied.update(
                 custom_controller_resources_config)
+        elif controller == Controllers.JOB_CONTROLLER:
+            # TODO(zhwu): Backward compatibility for the old config for managed
+            # job controller.
```
this is done iiuc?

cblmemo commented on tests/test_smoke.py:
```diff
@@ -61,9 +61,9 @@
 # To avoid the second smoke test reusing the cluster launched in the first
-# smoke test. Also required for test_spot_recovery to make sure the manual
+# smoke test. Also required for test_managed_jobs_recovery to make sure the manual
 # termination with aws ec2 does not accidentally terminate other spot clusters
```
```suggestion
# termination with aws ec2 does not accidentally terminate other managed job clusters
```

Michaelvll replied:
Fixed. Thanks!

cblmemo commented on tests/test_smoke.py:
```diff
@@ -61,9 +61,9 @@
+# smoke test. Also required for test_managed_jobs_recovery to make sure the manual
 # termination with aws ec2 does not accidentally terminate other spot clusters
-# from the different spot launch with the same cluster name but a different job
+# from the different job launch with the same cluster name but a different job
```
```suggestion
# from the different managed job launch with the same cluster name but a different job
```

Michaelvll replied:
Fixed. Thanks!

cblmemo commented on tests/test_smoke.py:
```diff
@@ -81,22 +81,22 @@
+    'to be ready"; sleep 5; s=$(sky jobs cancel -y -n {job_name}); '
     'done; echo "$s"; echo; echo; echo "$s"')
-# TODO(zhwu): make the spot controller on GCP.
+# TODO(zhwu): make the job controller on GCP.
```
what does this todo means

Michaelvll replied:
Fixed the comment. Thanks!

cblmemo commented on tests/test_smoke.py:
```diff
@@ -2220,38 +2220,31 @@ def test_stop_gcp_spot():
 
 
-# ---------- Testing managed spot ----------
-@pytest.mark.no_fluidstack  # FluidStack does not support spot instances
-@pytest.mark.no_azure  # Azure does not support spot instances
-@pytest.mark.no_lambda_cloud  # Lambda Cloud does not support spot instances
-@pytest.mark.no_ibm  # IBM Cloud does not support spot instances
-@pytest.mark.no_scp  # SCP does not support spot instances
-@pytest.mark.no_paperspace  # Papperspace does not support spot instances
-@pytest.mark.no_kubernetes  # Kubernetes does not have a notion of spot instances
```
we still need those since the YAML specified `use_spot: true`?

Michaelvll replied:
Good catch! It was a mistake introduced by a fix for a previous comment. Now changed it back. Thanks!

cblmemo reviewed (approved):
Thanks for this great work @Michaelvll !! It looks awesome to me. Left some nits and it should be ready to go! 🚀

Michaelvll commented on docs/source/images/job-dashboard.png:
We do not add the suffix for `On-Demand` only request for simplicity, and add the suffix for all other cases. Wdyt?

Michaelvll commented on docs/source/reference/config.rst:
```diff
@@ -224,7 +224,7 @@ Available fields and semantics:
     # GCP instances created by SkyPilot. They are used for accessing cloud
     # resources (e.g., private buckets) or launching new instances (e.g., for
-    # spot/serve controllers).
+    # job/serve controllers).
```
I was still calling it `job controller` because grammatically we normally do not use plural for an attributive noun, but after looking around, it seems fine to call it `jobs controller` as well.

Michaelvll commented on sky/cli.py:
```diff
@@ -816,31 +816,77 @@ def get_help(self, ctx):
+        if override_command_argument is not None:
+            overrides = []
+            for k, v in override_command_argument.items():
+                if isinstance(v, bool):
+                    overrides.append(f'--{k}')
```
Good point! Updated.

Michaelvll commented on sky/cli.py:
```diff
@@ -1219,63 +1265,79 @@ def exec(
+            if (record is None or
+                    record['status'] == status_lib.ClusterStatus.STOPPED):
+                msg = controller_type.value.default_hint_if_non_existent
+        except Exception:  # pylint: disable=broad-except
+            pass
```
Added a comment here.
